programs/dynamic-bonding-curve/src/instructions/ix_swap.rs

```diff
@@ -330,6 +384,108 @@ pub fn validate_single_swap_instruction<'c, 'info>(
             }
         }
     }
-
     Ok(())
 }
+
+struct ProcessSwapResult {
+    swap_result: SwapResult,
+    user_pay_input_amount: u64,
+    swap_in_parameters: SwapExactInParameters,
+}
+
+struct ProcessSwapParams<'a> {
+    pool: &'a mut VirtualPool,
+    config: &'a PoolConfig,
+    fee_mode: &'a FeeMode,
+    trade_direction: TradeDirection,
+    current_point: u64,
+    amount_0: u64,
+    amount_1: u64,
+    swap_mode: SwapMode,
+    rate_limiter: Option<&'a FeeRateLimiter>,
+}
+
+fn process_swap_exact_in(params: ProcessSwapParams<'_>) -> Result<ProcessSwapResult> {
+    let ProcessSwapParams {
+        amount_0: amount_in,
+        amount_1: minimum_amount_out,
+        pool,
+        config,
+        fee_mode,
+        trade_direction,
+        current_point,
+        swap_mode,
+        rate_limiter,
+    } = params;
+
+    let (swap_result, user_pay_input_amount) = pool.get_swap_exact_in_result(
+        config,
+        amount_in,
+        fee_mode,
+        trade_direction,
+        current_point,
+        swap_mode,
+        rate_limiter,
+    )?;
+
+    require!(
+        swap_result.output_amount >= minimum_amount_out,
+        PoolError::ExceededSlippage
+    );
+
+    Ok(ProcessSwapResult {
+        swap_result,
+        user_pay_input_amount,
+        swap_in_parameters: SwapExactInParameters {
+            amount_in,
+            minimum_amount_out,
+        },
+    })
+}
+
+fn process_swap_exact_out(params: ProcessSwapParams<'_>) -> Result<ProcessSwapResult> {
+    let ProcessSwapParams {
+        pool,
+        config,
+        fee_mode,
+        trade_direction,
+        current_point,
+        amount_0: amount_out,
+        amount_1: maximum_amount_in,
+        rate_limiter,
+        ..
+    } = params;
+
+    let swap_result = pool.get_swap_exact_out_result(
+        config,
+        amount_out,
+        fee_mode,
+        trade_direction,
+        current_point,
+        rate_limiter,
+    )?;
+
+    let included_fee_in_amount = if fee_mode.fees_on_input {
```

Comment:
we can move it to a function of struct SwapResult: `get_included_fee_in_amount(fees_on_input: bool)`